src/core/ddsi/src/ddsi_typelookup.c

```diff
@@ -0,0 +1,356 @@
+  }
+  GVTRACE (" tid "PTYPEIDFMT, PTYPEID (*tid));
+
+  ddsrt_mutex_lock (&gv->tl_admin_lock);
+  struct tl_meta *tlm = ddsi_tl_meta_lookup_locked (gv, tid);
+  if (tlm == NULL)
+  {
+    tlm = ddsrt_malloc (sizeof (*tlm));
+    memset (tlm, 0, sizeof (*tlm));
+    tlm->gv = gv;
+    tlm->state = TL_META_NEW;
+    tlm->type_id = ddsi_typeid_dup (tid);
+    ddsrt_hh_add (gv->tl_admin, tlm);
+    GVTRACE (" new %p", tlm);
+  }
+  if (src_ep_guid != NULL)
+  {
+    bool found = false;
+    for (uint32_t n = 0; !found && n < tlm->n_endpoints; n++)
+    {
+      if (!memcmp (&tlm->endpoints[n], src_ep_guid, sizeof (*src_ep_guid)))
+        found = true;
+    }
+    if (!found)
+    {
+      tlm->n_endpoints++;
+      tlm->endpoints = ddsrt_realloc (tlm->endpoints, tlm->n_endpoints * sizeof (*tlm->endpoints));
+      memcpy (&tlm->endpoints[tlm->n_endpoints - 1], src_ep_guid, sizeof (*src_ep_guid));
+      GVTRACE (" add ep "PGUIDFMT, PGUID (*src_ep_guid));
+    }
+  }
+  if (tlm->sertype == NULL && type != NULL)
+  {
+    tlm->sertype = ddsi_sertype_ref (type);
+    tlm->state = TL_META_RESOLVED;
+    GVTRACE (" resolved");
+  }
+  if (dst != NULL && guid_prefix_empty (&tlm->dst_prefix) && !guid_prefix_empty (dst))
+    memcpy (&tlm->dst_prefix, dst, sizeof (*dst));
+  tlm->refc++;
+  GVTRACE (" state %d refc %u\n", tlm->state, tlm->refc);
+  ddsrt_mutex_unlock (&gv->tl_admin_lock);
+  if (type_id == NULL)
+    ddsrt_free (tid);
+}
+
+void ddsi_tl_meta_unref (struct ddsi_domaingv *gv, const type_identifier_t *type_id, const ddsi_guid_t *ep_guid)
+{
+  ddsrt_mutex_lock (&gv->tl_admin_lock);
+  struct tl_meta *tlm = ddsi_tl_meta_lookup_locked (gv, type_id);
+  assert (tlm != NULL);
+  ddsi_tl_meta_unref_impl_locked (gv, tlm, ep_guid);
+  ddsrt_mutex_unlock (&gv->tl_admin_lock);
+}
+
+static struct participant *get_typelookup_pp (const struct ddsi_domaingv *gv, const ddsi_guid_prefix_t *guid_prefix)
+{
+  struct entidx_enum_participant est;
+  struct participant *pp = NULL;
+  bool pp_found = false;
+  entidx_enum_participant_init (&est, gv->entity_index);
+  while (!pp_found && (pp = entidx_enum_participant_next (&est)) != NULL)
+  {
+    if (guid_prefix_empty (guid_prefix) || memcmp (&pp->e.guid.prefix, guid_prefix, sizeof (pp->e.guid.prefix)) == 0)
+      pp_found = true;
+  }
+  entidx_enum_participant_fini (&est);
+  return pp;
+}
+
+bool ddsi_tl_request_type (struct ddsi_domaingv * const gv, const type_identifier_t *type_id)
+{
+  ddsrt_mutex_lock (&gv->tl_admin_lock);
+  struct tl_meta *tlm = ddsi_tl_meta_lookup_locked (gv, type_id);
+  if (tlm->state != TL_META_NEW)
+  {
+    GVTRACE ("ddsi_tl_request_type - state not-new (%u) for "PTYPEIDFMT"\n", tlm->state, PTYPEID (*type_id));
+    ddsrt_mutex_unlock (&gv->tl_admin_lock);
+    return false;
+  }
+
+  struct participant *pp = get_typelookup_pp (gv, &tlm->dst_prefix);
+  struct writer *wr = get_builtin_writer (pp, NN_ENTITYID_TL_SVC_BUILTIN_REQUEST_WRITER);
+  if (wr == NULL)
```

Comment:
So if an arbitrary participant was picked by `get_typelookup_pp` (the first one in the enumeration is a pretty arbitrary one, given how random the ids are) and it doesn't have a writer for requests, does that guarantee no other participant has a writer?